noir-projects/noir-contracts/contracts/protocol/tagging_contract/src/util.nr

```diff
@@ -0,0 +1,102 @@
+fn get_next_tag(
+    sender: AztecAddress,
+    recipient: AztecAddress,
+    hidden_sender: bool,
+    context: &mut PrivateContext,
+) -> Field {
+    if !handshake_exists(sender, recipient, hidden_sender) {
+        // The handshake doesn't exist, so we need to call the handshake function of the Handshaker contract.
+        context.call_private_function(
+            HANDSHAKER_CONTRACT_ADDRESS,
+            FunctionSelector::from_signature("handshake((Field,bool))"),
+            [sender.to_field(), recipient.to_field(), hidden_sender as Field],
+        );
+    }
```

Comment:
Introduced here after Nico's feedback. `handshake_exists` is an oracle that checks whether we've handshaked before and if not we will call the handshake function of the tagging contract.